worker/changestream/stream/stream.go

```diff
@@ -137,42 +157,95 @@ func (s *Stream) loop() error {
 
 			s.logger.Infof("Change stream has been unblocked")
 
-		case <-timer.Chan():
+		default:
 			changes, err := s.readChanges()
 			if err != nil {
+				// If the context was canceled, we're unsure if it's because
+				// the worker is dying, or if the context was canceled because
+				// the db was slow. In any case, continue and let the worker
+				// die if it's dying.
+				if errors.Is(errors.Cause(err), context.Canceled) {
+					continue
+				}
 				// If we get an error attempting to read the changes, the Txn
 				// will have retried multiple times. There just isn't anything
 				// we can do, so we just return an error.
 				return errors.Annotate(err, "reading change")
 			}
 
+			if len(changes) == 0 {
+				// The following uses the back-off strategy for polling the
+				// database for changes. If we repeatedly get no changes, then
+				// we back=off exponentially. This should prevent us from
+				// stuttering and should allow use to coalesce in the large
+				// case when nothing is happening.
+				attempt++
+				select {
+				case <-s.tomb.Dying():
+					return tomb.ErrDying
+				case <-s.clock.After(backOffStrategy(0, attempt)):
+				}
+				continue
+			}
+
+			// Reset the attempt counter if we get changes, so the back=off
+			// strategy is reset.
+			attempt = 0
+
+			// Term encapsulates a set of changes that are bounded by a
+			// coalesced set.
+			term := &Term{
+				done: make(chan struct{}),
+			}
+
+			traceEnabled := s.logger.IsTraceEnabled()
 			for _, change := range changes {
-				if s.logger.IsTraceEnabled() {
+				if traceEnabled {
 					s.logger.Tracef("change event: %v", change)
 				}
-				s.changes <- change
+				term.changes = append(term.changes, change)
 				s.lastID = change.id
 			}
 
-			// TODO (stickupkid): Adaptive polling based on the number of
-			// changes that are returned.
-			timer.Reset(PollInterval)
+			// Send the term to the terms channel, and wait for it to be
+			// completed. This will block the outer loop until the term has
+			// been completed. It is the responsibility of the consumer of the
+			// terms channel to ensure that the term is completed in the
+			// fastest possible time.
+			select {
+			case <-s.tomb.Dying():
+				return tomb.ErrDying
+			case s.terms <- term:
+			}
+
+			select {
```

Comment:
Is it worth a fail-safe here with a long time-out that returns with an error if the term is not getting marked done?

I think it's a good idea - we don't expect to encounter this case, but it just means we don't meekly play nice with bad downstream actors, rather we'll upend the cart if misused.

Reply:
Yes, I agree. I'm trying to understand the consequences of doing so. There are two cases here to solve:

  1. If we can't enqueue a term onto the stream terms channel.
  2. We're waiting for the term to complete.

If we can't enqueue, shows that the event multiplexer is struggling to read the terms channel. I think it might be worth implementing a back-off strategy to allow the event multiplexer to catch up.

If we're struggling to wait for a term to be done, we know that the term has been enqueued, but we don't actually know what subscription has been delivered with which term. In this case, I suspect it would be correct although painful to just bounce the stream, causing all downstream workers to bounce as well. We then know we've not got any subscriptions in a halfway house.